src/glb-director-xdp/main.go

```diff
@@ -342,6 +348,82 @@ func (app *Application) ReloadForwardingTable() {
 	}
 }
 
+func (app *Application) InitStatsCollection() {
+	if app.Config.StatsdPort != 0 {
+		client, err := statsd.New(fmt.Sprintf("127.0.0.1:%d", app.Config.StatsdPort))
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		client.Namespace = "glb.director."
+		client.Tags = append(client.Tags, "glb_engine:xdp")
+		app.StatsClient = client
+
+		go app.RunStatsCollection()
+	}
+}
+
+func diffAndSumGlobalStats(last []C.glb_global_stats, curr []C.glb_global_stats) C.glb_global_stats {
+	sum := C.glb_global_stats{}
+
+	for cpuIndex := 0; cpuIndex < len(last); cpuIndex++ {
+		sum.Processed += curr[cpuIndex].Processed - last[cpuIndex].Processed
+		sum.Encapsulated += curr[cpuIndex].Encapsulated - last[cpuIndex].Encapsulated
+
+		sum.UnknownFormat += curr[cpuIndex].UnknownFormat - last[cpuIndex].UnknownFormat
+		sum.NoMatchingBind += curr[cpuIndex].NoMatchingBind - last[cpuIndex].NoMatchingBind
+		sum.Matched += curr[cpuIndex].Matched - last[cpuIndex].Matched
+
+		sum.ErrorTable += curr[cpuIndex].ErrorTable - last[cpuIndex].ErrorTable
+		sum.ErrorSecret += curr[cpuIndex].ErrorSecret - last[cpuIndex].ErrorSecret
+		sum.ErrorHashConfig += curr[cpuIndex].ErrorHashConfig - last[cpuIndex].ErrorHashConfig
+		sum.ErrorMissingRow += curr[cpuIndex].ErrorMissingRow - last[cpuIndex].ErrorMissingRow
+		sum.ErrorCreatingSpace += curr[cpuIndex].ErrorCreatingSpace - last[cpuIndex].ErrorCreatingSpace
+		sum.ErrorMissingGatewayMAC += curr[cpuIndex].ErrorMissingGatewayMAC - last[cpuIndex].ErrorMissingGatewayMAC
+		sum.ErrorMissingSourceAddress += curr[cpuIndex].ErrorMissingSourceAddress - last[cpuIndex].ErrorMissingSourceAddress
+	}
+
+	return sum
+}
+
+func (app *Application) RunStatsCollection() {
+	globalCounters := app.Collection.Maps["glb_global_packet_counters"]
+	if globalCounters == nil {
+		log.Fatal("Could not load map glb_global_packet_counters")
```

Comment:
Probably don't want problems with stats collection to be fatal for everything else? Should this just print to error or something of that sort?

Reply:
The intention was that this would only ever fail at startup (where it's better to know something went wrong), but this ended up in a goroutine which technically makes no guarantee of timing. I'll move this out of the goroutine and into the initial setup to be explicit about that separation. The stats loop itself should definitely not cause failure.

The "fail on error" during init follows the pattern so far which is to fail during init prior to marking `READY=1` with systemd/rexec, so we don't degrade in some way and instead abort the start/reload with an error.

Reply:
Makes sense. Fatal in initial setup seems appropriate. 👍